EIPS/eip-7873.md

```diff
@@ -155,21 +155,41 @@ At the start of the block in which [EIP-3540](./eip-3540.md) activates, set the
 
 ```solidity
 {
-    /// Takes [tx_initcode_hash][salt][init_data] as input,
-    /// creates contract and returns the address or failure otherwise
+    /// Takes [tx_initcode_hash][salt][unsafe_flag][init_data] as input,
+    /// creates contract and returns the address or failure otherwise.
 
-    // init_data.length can be 0, but the first 2 words are mandatory
-    let size := calldatasize()
-    if lt(size, 64) { revert(0, 0) }
+    /// tx_initcode_hash and salt are 32 bytes wide, unsafe_flag is 1 byte wide, init_data can be any width.
+    /// If unsafe_flag is 0x00, `CALLER` is included in the salt and impacts the target address of the created account.
+    /// If unsafe_flag is non-zero, `CALLER` is excluded. Deployment by any sender will result in same target address. 
+    /// Deployments using non-zero unsafe_flag are susceptible to front-running.
 
-    calldatacopy(0, 0, size)  // copy tx_initcode_hash, salt and init_data to memory to hash
-    let final_salt := keccak256(0, size)
+    // init_data.length can be 0, but the first 65 bytes are mandatory
+    let size := calldatasize()
+    if lt(size, 65) { revert(0, 0) }
+
+    // copy tx_initcode_hash, salt, unsafe_flag and init_data to memory to hash.
+    // note that the unsafe_flag is included to prevent hash collisions between the two branches:
+    //   - unsafe_flag == 0, caller's address is appended by the code
+    //   - unsafe_flag == 1, caller's address is sent via init_data
+    calldatacopy(0, 0, size)
+
+    // mask out the 31 bytes that follow the flag in input data
+    let unsafe_flag := byte(0, calldataload(64))
```

Comment:
I had another thought on this - maybe it's better to require the unsafe_flag be just one lowest bit of that byte, so that the other bits can be possibly repurposed for other flags? The contract is not upgradable (nor planned to be re-deployed by state change), but... maybe this tiny amount forward compatibility would be useful?

This of course opens another can of worms: should we revert on higher bits set?

Reply:
The forward compatible option would be to force `unsafe_flag` to be `0x00` or `0x01` and revert on other values.

I also find that tempting. Using RJUMPV it would be very simple to express in the code. So I'd be in favor of this change.

Reply:
> The forward compatible option would be to force `unsafe_flag` to be `0x00` or `0x01` and revert on other values.
> 
> I also find that tempting. Using RJUMPV it would be very simple to express in the code. So I'd be in favor of this change.

Ah, right, if we revert on unsafe_flag > 0x01 then no need to worry about hashing. I find this an extra advantage of such restricted approach - otherwise unsafe_flag's higher 7 bits become de-facto salt, which is a bit weird.

@gumb0 can you elaborate on your reservations towards this (you've :-1:ed )?

Reply:
To me it feels like overthinking and unnecessary logic complication for speculative benefit. Current code with branching on `iszero(unsafe_flag)` looks fine, for hashing I would do Pawel's suggestion to hash 0 address instead of caller for unsafe mode and don't hash the flag.

Reply:
Resolved, see other thread and let's continue there